lib/all_futures/base.rb

```diff
@@ -3,26 +3,28 @@
 module AllFutures
   class Base < ActiveEntity::Base
     prepend ::AllFutures::Callbacks
-    include ActiveModel::Conversion
+    include ::ActiveModel::Conversion
+    include ::AllFutures::Persist
+    include ::AllFutures::Dirty
 
     def initialize(attributes = {})
-      super do
-        extend(::AllFutures::Persist)
-        extend(::AllFutures::Dirty)
-
-        @id ||= SecureRandom.uuid
-        @redis_key = "#{self.class.name}:#{@id}"
-        @new_record = !Kredis.redis.exists?(@redis_key)
-
-        @destroyed = false
-        @previously_new_record = false
-        @_trigger_destroy_callback = false
-        @_trigger_update_callback = false
-
-        @attributes.keys.each do |attr|
-          define_singleton_method("saved_change_to_#{attr}?") { saved_change_to_attribute?(attr) }
-          define_singleton_method("saved_change_to_#{attr}") { saved_change_to_attribute?(attr) ? [attribute_previously_was(attr), attribute_was(attr)] : nil }
-        end
+      # `active_entity/inheritance.rb:49` defaults `attributes` to `nil`, and our method signature has no effect
+      attributes ||= {}
+      
+      # in order to avoid FrozenError: can't modify id when persisted in `id=` 
+      attributes_for_super = attributes.key?(:id) ? attributes.except(:id) : attributes
+      super(attributes_for_super)
```

Comment:
So 2 conversation threads here: [1] why no more `super do` and [2] why the changes to deal with `id` in `attributes` (where there are 2 changes, this one and the `@id=` line). Let me address each:

1. I removed `super do` because at first I didn't understand what it did. I have since Googled and learned about parent class `initialize` methods having a `yield` that essentially provides a hook for the child class `initialize` method. So, we can put it back. However, removing the `attributes[:id]` stuff still leaves things broken, even with the `super do`

2. I understand and agree that `id` should be treated as a property and not an attribute, but the code path for `find` means that it is included in the `attributes` hash passed to initialize _sometimes_ (ln 41 `new json.merge(id: id)`). On that codepath, without the changes I made, the `id=` setter gets called and throws the error because the record isn't new if `:id` is in the `attributes` hash when passed to the parent class `initialize` method, or the found method could set a new `id` and disregard the passed in `id`. So, these two changes are intended to allow the `find` codepath to stay the same, but to treat `id` _not as an attribute_, but as a property, even when a persisted record is being initialised after a `find`

Reply:
Great explanation! Thank you for coming with me on this journey of discovery.